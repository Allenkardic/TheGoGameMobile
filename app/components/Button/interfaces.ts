import {ViewStyle, PressableProps, StyleProp} from 'react-native';

export type IButtonProps = React.FC<
  PressableProps & {
    onPress: any;
    isSubmitting?: boolean;
    hasError?: boolean;
    text: string;
    style?: StyleProp<ViewStyle>;
  }
>;
