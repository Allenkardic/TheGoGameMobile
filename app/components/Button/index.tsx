import React, {memo} from 'react';
import {StyleSheet, View, Pressable} from 'react-native';
import {IButtonProps} from './interfaces';
import Spinner from '../Spinner';
import Text from '../Typography';
import {Sizes, Colors} from '../../utils';

/**
 * This component is used to render a button.
 * @param props IButtonProps
 * @returns React.ReactElement
 */
const Button: IButtonProps = function Button({
  onPress,
  isSubmitting,
  text,
  style,
}) {
  return (
    <View style={[styles.container, style]}>
      <Pressable
        style={[styles.button]}
        onPress={onPress}
        disabled={isSubmitting || false}>
        {isSubmitting ? (
          <Spinner color="white" />
        ) : (
          <Text style={styles.btnText}>{text}</Text>
        )}
      </Pressable>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    borderRadius: Sizes.BUTTON_BORDER_RADIUS,
    overflow: 'hidden',
    borderWidth: 2,
    backgroundColor: Colors.secondary,
  },

  button: {
    height: Sizes.BUTTON_SIZE,
    paddingVertical: 2,
    justifyContent: 'center',
    borderRadius: Sizes.BUTTON_BORDER_RADIUS,
    alignItems: 'center',
  },

  btnText: {
    color: Colors.white,
  },
});

export default memo(Button);
