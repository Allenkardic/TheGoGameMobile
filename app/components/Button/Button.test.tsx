import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import Button from './';
import {jest} from '@jest/globals';

describe('Button Component', () => {
  let mockOnPress: jest.Mock;

  beforeEach(() => {
    jest.useFakeTimers(); // Mock timers
    mockOnPress = jest.fn();
  });

  afterEach(() => {
    jest.clearAllTimers(); // Clear any timers
    jest.useRealTimers(); // Reset to default behavior
  });

  it('renders correctly with default props', () => {
    const {toJSON} = render(
      <Button onPress={mockOnPress} text="Click Me" isSubmitting={false} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders spinner when isSubmitting is true', () => {
    const {toJSON} = render(
      <Button onPress={mockOnPress} text="Submit" isSubmitting={true} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('triggers onPress when button is pressed', () => {
    const {getByTestId} = render(
      <Button onPress={mockOnPress} text="Click Me" isSubmitting={false} />,
    );
    const button = getByTestId('button');
    fireEvent.press(button);
    expect(mockOnPress).toHaveBeenCalled();
  });

  it('applies custom styles', () => {
    const customStyle = {backgroundColor: 'blue'};
    const {toJSON} = render(
      <Button
        onPress={mockOnPress}
        text="Styled Button"
        isSubmitting={false}
        style={customStyle}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
