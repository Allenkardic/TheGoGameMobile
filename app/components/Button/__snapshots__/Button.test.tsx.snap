// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component applies custom styles 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#031a61",
        "borderRadius": 30,
        "borderWidth": 2,
        "overflow": "hidden",
      },
      {
        "backgroundColor": "blue",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 30,
          "height": 60,
          "justifyContent": "center",
          "paddingVertical": 2,
        },
      ]
    }
    testID="button"
  >
    <Text
      style={
        [
          {
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 24,
          },
          {
            "color": "#010a1a",
          },
          {
            "color": "#FFFFFF",
          },
          null,
          null,
        ]
      }
      testID="text"
    >
      Styled Button
    </Text>
  </View>
</View>
`;

exports[`Button Component renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#031a61",
        "borderRadius": 30,
        "borderWidth": 2,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 30,
          "height": 60,
          "justifyContent": "center",
          "paddingVertical": 2,
        },
      ]
    }
    testID="button"
  >
    <Text
      style={
        [
          {
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 24,
          },
          {
            "color": "#010a1a",
          },
          {
            "color": "#FFFFFF",
          },
          null,
          null,
        ]
      }
      testID="text"
    >
      Click Me
    </Text>
  </View>
</View>
`;

exports[`Button Component renders spinner when isSubmitting is true 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#031a61",
        "borderRadius": 30,
        "borderWidth": 2,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 30,
          "height": 60,
          "justifyContent": "center",
          "paddingVertical": 2,
        },
      ]
    }
    testID="button"
  >
    <ActivityIndicator
      color="white"
      size="small"
      style={
        [
          {
            "marginBottom": -4,
          },
          undefined,
        ]
      }
      testID="spinner"
    />
  </View>
</View>
`;
